### ISSUE TITLE
Feature request: multi touch panning

### ISSUE THREAD

**Issue by JarnoLeConte** [mapbox/mapbox-gl-js#2618]
When using a multi touch trackpad _on OSX_ I would let two finger slide gestures result in map panning instead of zooming. That gesture is very intuitive, and allows panning with inertia, which is easier to use than dragging the map, although both options should be available of course. Zooming could already be done by using the pinch gesture.

This is implemented by Apple as well in their default Maps application on OSX. It would be a great improvement in my opinion.

How to implement this? Could handlers be customised / overridden? Or should it be implemented in this library itself?

**Comment by 1ec5** [comment]
We have this behavior by default in the [Mapbox OS X SDK](https://github.com/mapbox/mapbox-gl-native/tree/master/platform/osx) for Cocoa applications, following MapKit’s lead. On the Web, Safari has a nonstandard [`GestureEvent`](https://developer.apple.com/library/iad/documentation/UserExperience/Reference/GestureEventClassReference/index.html) API to handle pinch and rotation gestures; I think it’s also used for two-finger panning gestures too.

**Comment by 1ec5** [comment]
On second thought, multitouch gestures are currently supported by GL JS in MobileSafari, in [touch_zoom_rotate.js](https://github.com/mapbox/mapbox-gl-js/blob/d6c34c81f7b0d6e77f1b25c2c080a3c5afba94ab/js/ui/handler/touch_zoom_rotate.js). Perhaps it could be generalized for desktop browsers too.

**Comment by JarnoLeConte** [comment]
Firstly I thought pinch is already implemented, but probably it just works because it's interpreted as scroll gesture and therefore zooms in/out. Now it makes sense why that gesture not working smoothly.

It would be great to generalise touch_zoom_rotate.js for desktop. I will take a look at it.

**Comment by jfirebaugh** [comment]
It looks like it isn't possible to detect pinch and rotate touch gestures on desktop browsers yet.

https://hammerjs.github.io/browser-support/

**Comment by 1ec5** [comment]
Desktop Safari isn’t included in that table; as I mentioned above, it does support a proprietary `GestureEvent` API for these gestures. Granted, desktop Safari has a lower usage share than some of the browsers listed there.

**Comment by JarnoLeConte** [comment]
What about the default scroll event to implement panning?

**Comment by 1ec5** [comment]
> It looks like it isn't possible to detect pinch and rotate touch gestures on desktop browsers yet.

On the desktop Internet Explorer, Edge, and Safari support these gestures via proprietary [`MSGestureEvent`](https://developer.mozilla.org/en-US/docs/Web/API/MSGestureEvent) and [`GestureEvent`](https://developer.mozilla.org/en-US/docs/Web/API/GestureEvent) APIs. I missed `MSGestureEvent` in https://github.com/mapbox/mapbox-gl-js/issues/2618#issuecomment-221308892.

> What about the default scroll event to implement panning?

That would leave no gesture to zoom the map. ArcGIS maps do this (gestures for scrolling but not zooming), but while it’s comfortable enough for multitouch mouse users, I find it maddening when trying to navigate the map with a multitouch trackpad.

**Comment by drott** [comment]
May I +1 this? Two finger gestures on mobile Safari and Chrome are separate into either a zoom intent or a drag intent. It'd be great to see this combined into zoom + pan gesture. 

Please also consider Chrome on Desktop / ChromeOS, laptop/desktop devices can have multi touch.

Thanks.

**Comment by mark-chief** [comment]
+ 1

**Comment by darrensq** [comment]
+1 on this

**Comment by tarciosaraiva** [comment]
+1

**Comment by mollymerp** [comment]
we would be glad to review a PR implementing this feature if anyone would like to contribute!

**Comment by pea** [comment]
I ended up writing this:

```
/**
 * Two Finger Mapbox Panning
 * new twoFingerMapboxPan(map)
 */
class twoFingerMapboxPan {
  constructor (mapbox) {
    this.mapbox = mapbox
    this.panStart = { x: 0, y: 0 }
    this.init()
  }
  init () {
    this.mapbox.getContainer().addEventListener('touchstart', this.touchStart.bind(this), false)
    this.mapbox.getContainer().addEventListener('touchmove', this.touchMove.bind(this), false)
    if ('ontouchstart' in document.documentElement) this.mapbox.dragPan.disable()
  }
  touchStart (event) {
    if (event.touches.length >= 2) {
      event.stopImmediatePropagation()
      event.preventDefault()

      let x = 0
      let y = 0

      for (let touch of Array.from(event.touches)) {
        x += touch.screenX
        y += touch.screenY
      }

      this.panStart.x = x / event.touches.length
      this.panStart.y = y / event.touches.length
    }
  }
  touchMove (event) {
    if (event.touches.length >= 2) {
      event.stopImmediatePropagation()
      event.preventDefault()

      let x = 0
      let y = 0

      for (let touch of Array.from(event.touches)) {
        x += touch.screenX
        y += touch.screenY
      }

      const movex = (x / event.touches.length) - this.panStart.x
      const movey = (y / event.touches.length) - this.panStart.y

      this.panStart.x = x / event.touches.length
      this.panStart.y = y / event.touches.length

      this.mapbox.panBy(
        [
          (movex * 1) / -1,
          (movey * 1) / -1
        ],
        { animate: false }
      )
    }
  }
}
```

Example: https://jsfiddle.net/kejxzpq0/3

**Comment by northkode** [comment]
@pea this solution is great for two finger pan but then completely disables pinch to zoom.

**Comment by pea** [comment]
Ah you're probably right. I think pinch zooming can be incorporated into the above code so I'll revisit it when I have time.

Would be great if all this could be included in Mapbox as standard though.

**Comment by northkode** [comment]
Sorry editing my message before seeing your reply. 
I fixed the solution

here it is 

```
// two-finger-pan.js
/**
 * Two Finger Mapbox Panning
 * new twoFingerMapboxPan(<mapbox instance>)
 */
const state = {
    mapbox: null,
    panStart: {
        x: 0,
        y: 0
    },
    scale: 1
}

export default (mapbox) => {
    state.mapbox = mapbox
    state.mapbox.getContainer().addEventListener('touchstart', touchStart, false)
    state.mapbox.getContainer().addEventListener('touchmove', touchMove, false)
    if ('ontouchstart' in document.documentElement) state.mapbox.dragPan.disable()
}

function touchStart(event) {
    if (event.touches.length === 2) {
        event.stopImmediatePropagation()
        event.preventDefault()

        let x = 0
        let y = 0

        for (let touch of Array.from(event.touches)) {
            x += touch.screenX
            y += touch.screenY
        }

        state.panStart.x = x / event.touches.length
        state.panStart.y = y / event.touches.length
    }
}

function touchMove(event, callback) {
    if (event.touches.length === 2) {
        if (state.scale == event.scale) {
            event.stopImmediatePropagation()
            event.preventDefault();
        }

        state.scale = event.scale;

        let x = 0
        let y = 0

        for (let touch of Array.from(event.touches)) {
            x += touch.screenX
            y += touch.screenY
        }

        const movex = (x / event.touches.length) - state.panStart.x
        const movey = (y / event.touches.length) - state.panStart.y

        state.panStart.x = x / event.touches.length
        state.panStart.y = y / event.touches.length

        state.mapbox.panBy(
            [
                (movex * 1) / -1,
                (movey * 1) / -1
            ], {
                animate: false
            }
        )
    }
}

```

**Comment by rantiev** [comment]
Instead of wrapping full clause block with if {} better do 

```
if (event.touches.length !== 2) {
  return
}

// here goes your piece that should work for several fingers logic handling
```

**Comment by msbarry** [comment]
What do people think about expanding the scope of the touch zoom rotate handler to include all 2 finger touch gestures? It could disambuguate between intent to drag, pan, rotate, and pitch. Ultimately it seems like we want to be able to move around the map like this demo lets you do: https://rawgit.com/axelpale/nudged/master/examples/nudged-gesture/index.html

**Comment by bravecow** [comment]
Code of @pea and @northkode as mapbox control on npm:

https://github.com/bravecow/mapbox-gl-multitouch

We use it in production. Thanks 👍

**Comment by jxlarrea** [comment]
@bravecow it's a good start but on desktop browsers (non touch) completely disables dragging and on mobile pinching no longer works.

This issue should really be a priority as the current behavior really breaks scrolling up and down a page with a map on mobile devices.

Google Maps behavior works great, Mapbox-GL should aim for that.

**Comment by bravecow** [comment]
@jxlarrea definitely you are right. this is just small workaround to allow users to scroll pages on mobile. we all are waiting for resolving this issue.

**Comment by jxlarrea** [comment]
@bravecow my workaround is a bit more dirty but in my opinion a bit more effective. Initially scrollZoom and dragPan are disabled. Once the user clicks/taps , zooms or use any of the controls of the map, these dragPan becomes enabled. 

This way, when the page loads the user is able to scroll through the page and only when the user clicks the map (which means they intent to interact with it), drag/pan becomes enabled. At least this way the map retains its pinch to zoom behavior.

 ```
  let map = new mapboxgl.Map({
            container: 'map',
            center: [center.lng, center.lat],
            zoom: 4,
            scrollZoom: false,
            dragPan: false,
            minZoom: 0,
            maxZoom: 11,
            attributionControl: false,
            fullScreenControl: true,
            style: 'mystyle'
        });
```
```
        let clickFunc = function (e) {
            map.dragPan.enable();
            map.off('click', clickFunc);
        };

        let zoomFunc = function (e) {
            if (e.source !== 'fitBounds') {
                map.dragPan.enable();
                map.off('zoomend', zoomFunc);
            }
        };

        map.on('click', clickFunc);
        map.on('zoomend', zoomFunc);
```

**Comment by woutervh-** [comment]
Here's my work-around:

```js
            this.map.on('dragstart', (event) => {
                if (event.originalEvent && 'touches' in event.originalEvent && event.originalEvent.touches.length >= 2) {
                    console.log('legit');
                } else {
                    if (this.map) {
                        this.map.dragPan.disable();
                        this.map.dragPan.enable();
                    }
                    console.log('not legit');
                }
            });
```

**Comment by maxmedina05** [comment]
> @pea this solution is great for two finger pan but then completely disables pinch to zoom.
> 
> i added in this
> 
> ```
> const state = {
>     mapbox: null,
>     panStart: {
>         x: 0,
>         y: 0
>     },
>     scale: 1
> }
> ```
> and then in touchMove
> 
> ```
> if (state.scale == event.scale) {
>             event.stopImmediatePropagation()
>             event.preventDefault();
>         }
> 
>         state.scale = event.scale;
> ```
> This allows for pinch to zoom and two finger panning

Only zoom in is working but it doesn't zoom out.
event.scale is always undefined.

**Comment by kikollan** [comment]
Any news on this? I am trying @pea solution, but I don't know how to install it (I am js newbie).

**Comment by Falkgeist** [comment]
I am using a map plugin for a CMS, referencing mapbox styles. This plugin doesn't allow to pan with two fingers (only) on mobile, so I thought I just go back to implementing Mapbox GL JS manually again, hoping this would be possible. Seems, here, it is the same issue.

Please, implement this, as I really don't want to keep using the ugly and not very customizable GMaps embeds, which seems the only way to not scroll-lock pages with maps on them.

Mapbox is of no use for responsive websites this way and – in my opinion – should make this feature a top priority for the devs.

**Comment by puremana** [comment]
Adding onto @jxlarrea 's solution, we only wanted this functionality to occur on mobile so we used
```js
        let mapData = {
            container: 'map',
            style: 'mapbox://styles/mapbox/streets-v11',
            center: [center.lng, center.lat],
            zoom: 6,
            scrollZoom: false,
            attributionControl: false,
            fullScreenControl: true
        };
        if (/Android|webOS|iPhone|iPad|iPod|BlackBerry/i.test(navigator.userAgent)) {
            mapData["dragPan"] = false;
        }
        
        var map = new mapboxgl.Map(mapData);

       // jxlarrea's code below...
```

**Comment by Akiyamka** [comment]
As workaround we can disable pan drag on one finger touch and re-enable when detected two finger touch. (inspired by [@woutervh- solution](https://github.com/mapbox/mapbox-gl-js/issues/2618#issuecomment-431819914))
```js
  const isTouchEvent = e => e.originalEvent && "touches" in e.originalEvent;
  const isTwoFingerTouch = e => e.originalEvent.touches.length >= 2;

  map.on("dragstart", event => {
    if (isTouchEvent(event) && !isTwoFingerTouch(event)) {
       map.dragPan.disable();
    }
  });

  // Drag events not emited after dragPan disabled, so I use touch event here
  map.on("touchstart", event => {
      if (isTouchEvent(event) && isTwoFingerTouch(event)) {
       map.dragPan.enable();
    }
  });
```
Sandbox:
https://codesandbox.io/s/react-mapbox-gl-cdzbz

**Comment by betabong** [comment]
My workaround (allows scrolling via 1 finger touch):

`map.dragPan.disable();
map.scrollZoom.disable();

map.on('touchstart', event => {
    const e = event.originalEvent;
    if (e && 'touches' in e) {
        if (e.touches.length > 1) {
            this.map.dragPan.enable();
        } else {
            this.map.dragPan.disable();
        }
    }
});`

**Comment by fabiocaccamo** [comment]
It would be nice to have a simple option to do this.

**Comment by DeBraid** [comment]
For future searches that lead here (I am using `react-map-gl` so be aware.

To make scrolling better on mobile, set `dragPan` and `touchAction` like so:

```js
<ReactMapGL
      mapboxApiAccessToken={TOKEN}
      mapStyle={mapStyle}
      dragPan={false} // <-- 
      touchAction={'pan-y'} // <-- 
    >
```

> By default, the map captures all touch interactions. `touchAction` prop is useful for mobile applications to unblock default scrolling behaviour. For example, use the combination `dragPan: false` and `touchAction: 'pan-y'` to allow vertical page scroll when dragging over the map.

Docs: https://visgl.github.io/react-map-gl/docs/api-reference/interactive-map

**Comment by fabiocaccamo** [comment]
I solved the mobile scroll issue in this way:
```javascript
if (/Android|webOS|iPhone|iPad|iPod|BlackBerry/i.test(navigator.userAgent)) {
    // https://github.com/mapbox/mapbox-gl-js/issues/2618
    var map = this._map;
    map.dragPan.disable();
    map.scrollZoom.disable();
    map.touchPitch.disable()
    map.on('touchstart', function(e) {
        var oe = e.originalEvent;
        if (oe && 'touches' in oe) {
            if (oe.touches.length > 1) {
                oe.stopImmediatePropagation();
                map.dragPan.enable();
            } else {
                map.dragPan.disable();
            }
        }
    });
}
```

**Comment by jdato** [comment]
+1

**Comment by Jonatan-r** [comment]
+1

**Comment by ygoe** [comment]
The original comment talks about multi-touch trackpads on a Mac. I know nothing about these. My computer has a mouse. But my smartphone only has a touch-screen and it must be used to scroll the page. Now with a map in fullscreen view, the **page cannot be scrolled anymore!** Only the map can be panned. Is that what this issue is about? For me this is a major bug and it has been solved years ago in Google Maps. Is anybody even looking at this?

**Comment by ygoe** [comment]
None of the workarounds presented here work satisfyingly. It just won't feel solid.

Hours later I'm done with my solution inspired by Google Maps (but not the same because that would require work inside mapbox). At first, I disable all map surface interactions completely to not interfer with page scrolling in any way. When a touchmove or wheel event occurs, a message is displayed, instructing the user to (de)activate the map with a double-click. Since the dblclick event is massively broken here, I detect two quick click events (Firefox still fires 3, the third is discarded to not deactive the map again on a triple-click). While activated, all map interactions are fully active. The map's control buttons remain active at any time. Double-click zoom is always disabled, I don't need that. This is a somewhat manual user experience, but at least it's explicit and predictable.

I might try and extend this to enlarge the map to the entire viewport while activated. Maybe. That could make the current state more clearly visible.

This is the code, I hope it helps anybody who is just as frustrated as me about this unresponsive map control. At least the rest of it works very nicely.

```js
// Create the map
const map = new mapboxgl.Map({
    container: mapContainer,
    style: "mapbox://styles/mapbox/streets-v11",
    doubleClickZoom: false,
    scrollZoom: false,
    touchZoomRotate: false,
    touchPitch: false,
    dragPan: false,
    dragRotate: false
});
map.touchZoomRotate.disableRotation();
map.addControl(new mapboxgl.NavigationControl({
    showCompass: false
}));

// Create a message that instructs the user to double-click/tap to activate the map control
let mapActive = false;
const mapMessage = document.createElement("div");
mapMessage.classList.add("map-message");
mapMessage.style.height = mapContainer.style.height;   // NOTE: mapContainer is the map element; redo this on resize
mapMessage.style.marginTop = "-" + mapContainer.style.height;
mapMessage.style.pointerEvents = "none";
mapMessage.style.position = "relative";   // required for z-index
mapMessage.style.zIndex = "5";   // required to overlay map controls, too (z-index: 2)
mapMessage.style.opacity = "0";
mapMessage.style.background = "rgba(0, 0, 0, 0.4)";
mapMessage.style.display = "flex";
mapMessage.style.justifyContent = "center";
mapMessage.style.alignItems = "center";
mapMessage.style.textAlign = "center";
mapMessage.style.color = "white";
mapMessage.style.fontSize = "1.4em";
mapMessage.style.lineHeight = "130%";
mapMessage.style.padding = "1rem";
mapMessage.style.transition = "opacity 0.5s";
mapMessage.textContent = "(De)activate the map with a double-click";
innerContainer.appendChild(mapMessage);   // NOTE: innerContainer is the parent of the map element; add this directly after mapContainer
let mapMessageTimeout;
const showMessage = () => {
    mapMessage.style.opacity = "0.999";   // opacity 1 makes it invisible!
    clearTimeout(mapMessageTimeout);
    mapMessageTimeout = setTimeout(() => {
        mapMessage.style.opacity = "0";
    }, 1500);
};
map.on("touchmove", e => {
    if (e.originalEvent.touches && e.originalEvent.touches.length === 1 && !mapActive)
        showMessage();
});
map.on("wheel", e => {
    if (!mapActive)
        showMessage();
});

// Activate or deactivate the map control with a double click/tap to prevent unintended
// scrolling, panning or zooming on the usual page scroll interactions.
// The dblclick event is massively broken in Firefox (fires twice, for mouse and touch)
// and Chrome (not at all), so we have to do that manually.
let lastClickTime;
map.on("click", event => {
    if (event.originalEvent.detail > 2)
        return;   // Ignore triple-click in Firefox (no event in Chrome)
    const now = Date.now();
    if (lastClickTime >= now - 500) {
        //console.log("double click", event);
        mapActive = !mapActive;
        if (mapActive) {
            map.scrollZoom.enable();
            map.touchZoomRotate.enable();
            map.dragPan.enable();
            mapContainer.style.outline = "2px solid limegreen";   // visualisation for the activated map
        }
        else {
            map.scrollZoom.disable();
            map.touchZoomRotate.disable();
            map.dragPan.disable();
            mapContainer.style.outline = "none";
        }
        lastClickTime = undefined;
    }
    else {
        lastClickTime = now;
    }
});
```

Tested with Firefox 87 (Windows), Chrome 89 (Windows), Brave (Android), Firefox (Android) and Opera (Android).

**Comment by ygoe** [comment]
I just updated my previous code and added a workaround for #10593 (wrong mouse cursor displayed when dragPan is disabled).

**Comment by jxlarrea** [comment]
@ygoe amazing! Thanks for the code! Now the map behavior is almost perfect and works just like Google Maps.

**Comment by ansis** [comment]
While the first comment was talking about something slightly different, most of the discussion has been about requiring two touches to pan the map in order to avoid capturing page scrolls. This was added in https://github.com/mapbox/mapbox-gl-js/pull/11116 and is going to be part of the v2.6.0 release. We've just released the beta, which can be tested with either by pulling in `2.6.0-beta.1` from npm or by using:
- `https://api.mapbox.com/mapbox-gl-js/v2.6.0-beta.1/mapbox-gl.js`
- `https://api.mapbox.com/mapbox-gl-js/v2.6.0-beta.1/mapbox-gl.css` (be sure to update this as well!)

If you set the `new Map({ cooperativeGestures: true, ... })` then two touches will be required to pan the map. On desktop, `ctrl` or `⌘` will be required to scroll to avoid a similar problem.



![Screen Shot 2021-10-20 at 9 15 40 PM](https://user-images.githubusercontent.com/1421652/138193809-257d4fe5-55cd-40a1-88ee-6a81d893536a.png)




If you notice any problems, please open a new issue!


release notes: https://github.com/mapbox/mapbox-gl-js/releases/tag/v2.6.0-beta.1
